{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot gene set comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, we want to compare prediction of mutations in the genes from our cancer gene set, derived from [Vogelstein et al. 2013](https://science.sciencemag.org/content/339/6127/1546), with two other sets of potentially relevant genes. These are:\n",
    "\n",
    "* The most frequently mutated genes in TCGA\n",
    "* A set of random genes in TCGA, that meet our mutation count threshold for 2 or more cancer types\n",
    " \n",
    "We selected enough genes in each of these gene sets to (approximately) match the count of the Vogelstein et al. gene set.\n",
    " \n",
    "In these experiments we only used gene expression data, and we used the set of TCGA samples that have both gene expression and MC3 somatic mutation data. The files analyzed in this notebook were generated by the `run_mutation_prediction.py` script.\n",
    "\n",
    "Notebook parameters:\n",
    "* SIG_ALPHA (float): significance cutoff (after FDR correction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from adjustText import adjust_text\n",
    "\n",
    "import mpmp.config as cfg\n",
    "import mpmp.utilities.analysis_utilities as au\n",
    "import mpmp.utilities.plot_utilities as plu"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set results directory\n",
    "vogelstein_results_dir = Path(cfg.results_dirs['mutation'],\n",
    "                              'vogelstein_expression_only',\n",
    "                              'gene').resolve()\n",
    "top_50_results_dir = Path(cfg.results_dirs['mutation'],\n",
    "                          'top_50_expression_only',\n",
    "                          'gene').resolve()\n",
    "random_50_results_dir = Path(cfg.results_dirs['mutation'],\n",
    "                             '50_random_expression_only',\n",
    "                             'gene').resolve()\n",
    "\n",
    "# set significance cutoff after FDR correction\n",
    "SIG_ALPHA = 0.001\n",
    "\n",
    "# if True, save figures to ./images directory\n",
    "SAVE_FIGS = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4080, 10)\n",
      "[42  1]\n",
      "['expression']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>auroc</th>\n",
       "      <th>aupr</th>\n",
       "      <th>identifier</th>\n",
       "      <th>training_data</th>\n",
       "      <th>signal</th>\n",
       "      <th>seed</th>\n",
       "      <th>data_type</th>\n",
       "      <th>fold</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.99980</td>\n",
       "      <td>0.99808</td>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>expression</td>\n",
       "      <td>signal</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>vogelstein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.75084</td>\n",
       "      <td>0.43962</td>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>expression</td>\n",
       "      <td>signal</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>vogelstein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.68143</td>\n",
       "      <td>0.30787</td>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>expression</td>\n",
       "      <td>signal</td>\n",
       "      <td>42</td>\n",
       "      <td>cv</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>vogelstein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.99999</td>\n",
       "      <td>0.99989</td>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>expression</td>\n",
       "      <td>signal</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>vogelstein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.66563</td>\n",
       "      <td>0.43830</td>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>expression</td>\n",
       "      <td>signal</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>vogelstein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     auroc     aupr identifier training_data  signal  seed data_type  fold  \\\n",
       "0  0.99980  0.99808     MAP3K1    expression  signal    42     train     0   \n",
       "1  0.75084  0.43962     MAP3K1    expression  signal    42      test     0   \n",
       "2  0.68143  0.30787     MAP3K1    expression  signal    42        cv     0   \n",
       "3  0.99999  0.99989     MAP3K1    expression  signal    42     train     1   \n",
       "4  0.66563  0.43830     MAP3K1    expression  signal    42      test     1   \n",
       "\n",
       "  experiment    gene_set  \n",
       "0       gene  vogelstein  \n",
       "1       gene  vogelstein  \n",
       "2       gene  vogelstein  \n",
       "3       gene  vogelstein  \n",
       "4       gene  vogelstein  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load raw data\n",
    "vogelstein_df = au.load_stratified_prediction_results(vogelstein_results_dir, 'gene')\n",
    "vogelstein_df = vogelstein_df[vogelstein_df.training_data.isin(['expression'])]\n",
    "vogelstein_df['gene_set'] = 'vogelstein'\n",
    "\n",
    "# make sure that we're correctly pointing to raw data for non-methylation data types\n",
    "# and that we have data for two replicates (two random seeds)\n",
    "print(vogelstein_df.shape)\n",
    "print(vogelstein_df.seed.unique())\n",
    "print(vogelstein_df.training_data.unique())\n",
    "vogelstein_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4080, 10)\n",
      "[42  1]\n",
      "['expression']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>auroc</th>\n",
       "      <th>aupr</th>\n",
       "      <th>identifier</th>\n",
       "      <th>training_data</th>\n",
       "      <th>signal</th>\n",
       "      <th>seed</th>\n",
       "      <th>data_type</th>\n",
       "      <th>fold</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.98433</td>\n",
       "      <td>0.98653</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>top_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.50011</td>\n",
       "      <td>0.48180</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>top_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.52088</td>\n",
       "      <td>0.49482</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>cv</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>top_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.97987</td>\n",
       "      <td>0.97430</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>top_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.50415</td>\n",
       "      <td>0.49414</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>top_50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     auroc     aupr identifier training_data    signal  seed data_type  fold  \\\n",
       "0  0.98433  0.98653       TP53    expression  shuffled    42     train     0   \n",
       "1  0.50011  0.48180       TP53    expression  shuffled    42      test     0   \n",
       "2  0.52088  0.49482       TP53    expression  shuffled    42        cv     0   \n",
       "3  0.97987  0.97430       TP53    expression  shuffled    42     train     1   \n",
       "4  0.50415  0.49414       TP53    expression  shuffled    42      test     1   \n",
       "\n",
       "  experiment gene_set  \n",
       "0       gene   top_50  \n",
       "1       gene   top_50  \n",
       "2       gene   top_50  \n",
       "3       gene   top_50  \n",
       "4       gene   top_50  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load raw data\n",
    "top_50_df = au.load_stratified_prediction_results(top_50_results_dir, 'gene')\n",
    "top_50_df = top_50_df[top_50_df.training_data.isin(['expression'])]\n",
    "top_50_df['gene_set'] = 'top_50'\n",
    "\n",
    "# make sure that we're correctly pointing to raw data for non-methylation data types\n",
    "# and that we have data for two replicates (two random seeds)\n",
    "print(top_50_df.shape)\n",
    "print(top_50_df.seed.unique())\n",
    "print(top_50_df.training_data.unique())\n",
    "top_50_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4080, 10)\n",
      "[42  1]\n",
      "['expression']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>auroc</th>\n",
       "      <th>aupr</th>\n",
       "      <th>identifier</th>\n",
       "      <th>training_data</th>\n",
       "      <th>signal</th>\n",
       "      <th>seed</th>\n",
       "      <th>data_type</th>\n",
       "      <th>fold</th>\n",
       "      <th>experiment</th>\n",
       "      <th>gene_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.98433</td>\n",
       "      <td>0.98653</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>random_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.50011</td>\n",
       "      <td>0.48180</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>random_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.52088</td>\n",
       "      <td>0.49482</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>cv</td>\n",
       "      <td>0</td>\n",
       "      <td>gene</td>\n",
       "      <td>random_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.97987</td>\n",
       "      <td>0.97430</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>train</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>random_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.50415</td>\n",
       "      <td>0.49414</td>\n",
       "      <td>TP53</td>\n",
       "      <td>expression</td>\n",
       "      <td>shuffled</td>\n",
       "      <td>42</td>\n",
       "      <td>test</td>\n",
       "      <td>1</td>\n",
       "      <td>gene</td>\n",
       "      <td>random_50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     auroc     aupr identifier training_data    signal  seed data_type  fold  \\\n",
       "0  0.98433  0.98653       TP53    expression  shuffled    42     train     0   \n",
       "1  0.50011  0.48180       TP53    expression  shuffled    42      test     0   \n",
       "2  0.52088  0.49482       TP53    expression  shuffled    42        cv     0   \n",
       "3  0.97987  0.97430       TP53    expression  shuffled    42     train     1   \n",
       "4  0.50415  0.49414       TP53    expression  shuffled    42      test     1   \n",
       "\n",
       "  experiment   gene_set  \n",
       "0       gene  random_50  \n",
       "1       gene  random_50  \n",
       "2       gene  random_50  \n",
       "3       gene  random_50  \n",
       "4       gene  random_50  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load raw data\n",
    "random_50_df = au.load_stratified_prediction_results(random_50_results_dir, 'gene')\n",
    "random_50_df = random_50_df[random_50_df.training_data.isin(['expression'])]\n",
    "random_50_df['gene_set'] = 'random_50'\n",
    "\n",
    "# make sure that we're correctly pointing to raw data for non-methylation data types\n",
    "# and that we have data for two replicates (two random seeds)\n",
    "print(random_50_df.shape)\n",
    "print(random_50_df.seed.unique())\n",
    "print(random_50_df.training_data.unique())\n",
    "random_50_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>delta_mean</th>\n",
       "      <th>p_value</th>\n",
       "      <th>corr_pval</th>\n",
       "      <th>reject_null</th>\n",
       "      <th>training_data</th>\n",
       "      <th>nlog10_p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>TP53</td>\n",
       "      <td>0.418908</td>\n",
       "      <td>3.030301e-13</td>\n",
       "      <td>2.575756e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>10.589095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>TP53</td>\n",
       "      <td>0.418908</td>\n",
       "      <td>3.030301e-13</td>\n",
       "      <td>2.575756e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>top_50</td>\n",
       "      <td>10.589095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>TP53</td>\n",
       "      <td>0.418908</td>\n",
       "      <td>3.030301e-13</td>\n",
       "      <td>2.575756e-11</td>\n",
       "      <td>True</td>\n",
       "      <td>random_50</td>\n",
       "      <td>10.589095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>0.603976</td>\n",
       "      <td>6.247747e-11</td>\n",
       "      <td>2.655293e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>8.575888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>0.387666</td>\n",
       "      <td>1.552123e-10</td>\n",
       "      <td>4.192733e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>8.377503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>APC</td>\n",
       "      <td>0.593855</td>\n",
       "      <td>1.973051e-10</td>\n",
       "      <td>4.192733e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>8.377503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>APC</td>\n",
       "      <td>0.593855</td>\n",
       "      <td>1.973051e-10</td>\n",
       "      <td>8.385466e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>top_50</td>\n",
       "      <td>8.076473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>0.576172</td>\n",
       "      <td>1.579724e-09</td>\n",
       "      <td>2.685531e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>7.570970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>0.576172</td>\n",
       "      <td>1.579724e-09</td>\n",
       "      <td>4.121261e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>top_50</td>\n",
       "      <td>7.384970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>0.468161</td>\n",
       "      <td>1.939417e-09</td>\n",
       "      <td>2.747507e-08</td>\n",
       "      <td>True</td>\n",
       "      <td>vogelstein</td>\n",
       "      <td>7.561061</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  delta_mean       p_value     corr_pval  reject_null training_data  \\\n",
       "82    TP53    0.418908  3.030301e-13  2.575756e-11         True    vogelstein   \n",
       "77    TP53    0.418908  3.030301e-13  2.575756e-11         True        top_50   \n",
       "77    TP53    0.418908  3.030301e-13  2.575756e-11         True     random_50   \n",
       "23    EGFR    0.603976  6.247747e-11  2.655293e-09         True    vogelstein   \n",
       "17  CDKN2A    0.387666  1.552123e-10  4.192733e-09         True    vogelstein   \n",
       "2      APC    0.593855  1.973051e-10  4.192733e-09         True    vogelstein   \n",
       "5      APC    0.593855  1.973051e-10  8.385466e-09         True        top_50   \n",
       "69    PTEN    0.576172  1.579724e-09  2.685531e-08         True    vogelstein   \n",
       "64    PTEN    0.576172  1.579724e-09  4.121261e-08         True        top_50   \n",
       "51    KRAS    0.468161  1.939417e-09  2.747507e-08         True    vogelstein   \n",
       "\n",
       "     nlog10_p  \n",
       "82  10.589095  \n",
       "77  10.589095  \n",
       "77  10.589095  \n",
       "23   8.575888  \n",
       "17   8.377503  \n",
       "2    8.377503  \n",
       "5    8.076473  \n",
       "69   7.570970  \n",
       "64   7.384970  \n",
       "51   7.561061  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combine results dataframes\n",
    "results_df = (\n",
    "    pd.concat((vogelstein_df, top_50_df, random_50_df))\n",
    "      .drop(columns=['training_data', 'experiment'])\n",
    "      .rename(columns={'gene_set': 'training_data'})\n",
    ")\n",
    "\n",
    "all_results_df = au.compare_all_data_types(results_df,\n",
    "                                           SIG_ALPHA,\n",
    "                                           filter_genes=False)\n",
    "\n",
    "all_results_df['nlog10_p'] = -np.log10(all_results_df.corr_pval)\n",
    "all_results_df.sort_values(by='p_value').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "random: 12/85\n",
      "most mutated: 15/85\n"
     ]
    }
   ],
   "source": [
    "sns.set({'figure.figsize': (24, 6)})\n",
    "sns.set_style('whitegrid')\n",
    "fig, axarr = plt.subplots(1, 3)\n",
    "\n",
    "gene_set_map = {\n",
    "    'random_50': 'random',\n",
    "    'top_50': 'most mutated',\n",
    "    'vogelstein': 'Vogelstein et al.'\n",
    "}\n",
    "all_results_df.training_data.replace(to_replace=gene_set_map, inplace=True)\n",
    "\n",
    "plu.plot_volcano_baseline(all_results_df,\n",
    "                          axarr,\n",
    "                          gene_set_map,\n",
    "                          SIG_ALPHA,\n",
    "                          verbose=True,\n",
    "                          color_overlap=True)\n",
    "\n",
    "if SAVE_FIGS:\n",
    "    images_dir = Path(cfg.images_dirs['mutation'])\n",
    "    images_dir.mkdir(exist_ok=True)\n",
    "    plt.savefig(images_dir / 'expression_vs_shuffled.svg', bbox_inches='tight')\n",
    "    plt.savefig(images_dir / 'expression_vs_shuffled.png',\n",
    "                dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set({'figure.figsize': (8, 6)})\n",
    "sns.set_style('whitegrid')\n",
    "fig, axarr = plt.subplots(1, 1)\n",
    "\n",
    "gene_set_map = {\n",
    "    'vogelstein': 'Vogelstein et al.'\n",
    "}\n",
    "all_results_df.training_data.replace(to_replace=gene_set_map, inplace=True)\n",
    "\n",
    "plu.plot_volcano_baseline(all_results_df,\n",
    "                          axarr,\n",
    "                          gene_set_map,\n",
    "                          SIG_ALPHA,\n",
    "                          verbose=True)\n",
    "\n",
    "if SAVE_FIGS:\n",
    "    images_dir = Path(cfg.images_dirs['mutation'])\n",
    "    images_dir.mkdir(exist_ok=True)\n",
    "    plt.savefig(images_dir / 'expression_vogelstein.svg', bbox_inches='tight')\n",
    "    plt.savefig(images_dir / 'expression_vogelstein.png',\n",
    "                dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set({'figure.figsize': (10, 6)})\n",
    "sns.set_style('whitegrid')\n",
    "# we want these colors to be different than the expression/methylation ones\n",
    "sns.set_palette('Set2')\n",
    "fig, axarr = plt.subplots(1, 1)\n",
    "\n",
    "all_results_df = (all_results_df\n",
    "   .rename(columns={'training_data': 'gene_set'})\n",
    "   .sort_values(by='gene_set', ascending=False)\n",
    ")\n",
    "\n",
    "# plot mean performance over all genes in Vogelstein dataset\n",
    "ax = axarr\n",
    "sns.boxplot(data=all_results_df, x='gene_set', y='delta_mean',\n",
    "            notch=True, ax=ax)\n",
    "ax.set_title('Performance distribution for all genes in each gene set, gene expression only', size=14)\n",
    "ax.set_xlabel('Target gene set', size=13)\n",
    "ax.set_ylabel('AUPR(signal) - AUPR(shuffled)', size=13)\n",
    "ax.set_ylim(-0.2, max(all_results_df.delta_mean + 0.05))\n",
    "for tick in ax.get_xticklabels():\n",
    "    tick.set_fontsize(13)\n",
    "    \n",
    "plt.tight_layout()\n",
    "\n",
    "if SAVE_FIGS:\n",
    "    plt.savefig(images_dir / 'expression_boxes.svg', bbox_inches='tight')\n",
    "    plt.savefig(images_dir / 'expression_boxes.png',\n",
    "                dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set({'figure.figsize': (18, 6)})\n",
    "sns.set_style('whitegrid')\n",
    "# we want these colors to be different than the expression/methylation ones\n",
    "sns.set_palette('Set2')\n",
    "fig, axarr = plt.subplots(1, 2)\n",
    "\n",
    "all_results_df.replace({'gene_set': gene_set_map}, inplace=True)\n",
    "\n",
    "# plot mean performance over all genes in Vogelstein dataset\n",
    "ax = axarr[0]\n",
    "sns.stripplot(data=all_results_df, x='gene_set', y='delta_mean', ax=ax)\n",
    "ax.set_title('Prediction for all genes, performance vs. gene set')\n",
    "ax.set_xlabel('Target gene set')\n",
    "ax.set_ylabel('AUPR(signal) - AUPR(shuffled)')\n",
    "ax.set_ylim(-0.2, max(all_results_df.delta_mean + 0.05))\n",
    "\n",
    "# plot mean performance for genes that are significant for at least one data type\n",
    "ax = axarr[1]\n",
    "gene_list = all_results_df[all_results_df.reject_null == True].gene.unique()\n",
    "print(gene_list.shape)\n",
    "print(gene_list)\n",
    "sns.stripplot(data=all_results_df[all_results_df.gene.isin(gene_list)],\n",
    "              x='gene_set', y='delta_mean', ax=ax)\n",
    "ax.set_title('Prediction for significant genes, performance vs. gene set')\n",
    "ax.set_xlabel('Target gene set')\n",
    "ax.set_ylabel('AUPR(signal) - AUPR(shuffled)')\n",
    "ax.set_ylim(-0.2, max(all_results_df.delta_mean + 0.05))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate gene set overlap\n",
    " \n",
    "Of the significantly predictable genes in the top/random gene sets, how many of them are in the Vogelstein gene set?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from venn import venn\n",
    "\n",
    "# first look at overlap of all genes\n",
    "genes_in_gene_set = {}\n",
    "for gene_set in all_results_df.gene_set.unique():\n",
    "    gene_list = all_results_df[all_results_df.gene_set == gene_set].gene.unique()\n",
    "    print(gene_set, len(gene_list))\n",
    "    genes_in_gene_set[gene_set] = set(gene_list)\n",
    "\n",
    "venn(genes_in_gene_set)\n",
    "plt.title('Gene overlap between all genes in gene set', size=14)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now look at overlap of significant genes\n",
    "genes_in_gene_set = {}\n",
    "for gene_set in all_results_df.gene_set.unique():\n",
    "    gene_list = all_results_df[(all_results_df.gene_set == gene_set) &\n",
    "                               (all_results_df.reject_null)].gene.unique()\n",
    "    print(gene_set, len(gene_list))\n",
    "    genes_in_gene_set[gene_set] = set(gene_list)\n",
    "\n",
    "venn(genes_in_gene_set)\n",
    "plt.title('Gene overlap between significantly predictable genes in gene set', size=14)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mpmp]",
   "language": "python",
   "name": "conda-env-mpmp-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
