{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download cancer gene sets\n",
    "\n",
    "We want to download the set of cancer-associated genes from the [COSMIC Cancer Gene Census](https://cancer.sanger.ac.uk/cosmic/census), and from [Bailey et al. 2018](https://www.sciencedirect.com/science/article/pii/S009286741830237X), in order to use these genes in our experiments as a comparison/complement to the Vogelstein et al. gene set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import mpmp.config as cfg\n",
    "import mpmp.utilities.data_utilities as du\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download COSMIC CGC data\n",
    "\n",
    "We downloaded the original CGC data directly from the Sanger Institute website linked above - you need to create an account there to download the .tsv file, so we can't do it programmatically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(445, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>12.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10006,ABI-1,ABI1,E3B1,ENSG00000136754.17,Q8IZP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25,ABL,ABL1,ENSG00000097007.17,JTK7,P00519,c-A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL2</th>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27</td>\n",
       "      <td>1:179099327-179229601</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27,ABL2,ABLL,ARG,ENSG00000143322.19,P42684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR3</th>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007</td>\n",
       "      <td>2:236569641-236582358</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>37.30</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57007,ACKR3,CMKOR1,CXCR7,ENSG00000144476.5,GPR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1</th>\n",
       "      <td>activin A receptor, type I</td>\n",
       "      <td>90</td>\n",
       "      <td>2:157736444-157875111</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>24.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DIPG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>Fibrodysplasia ossificans progressiva</td>\n",
       "      <td>90,ACVR1,ACVR1A,ACVRLK2,ALK2,ENSG00000115170.1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          Name  Entrez GeneId  \\\n",
       "Gene Symbol                                                                     \n",
       "ABI1                                          abl-interactor 1          10006   \n",
       "ABL1         v-abl Abelson murine leukemia viral oncogene h...             25   \n",
       "ABL2            c-abl oncogene 2, non-receptor tyrosine kinase             27   \n",
       "ACKR3                            atypical chemokine receptor 3          57007   \n",
       "ACVR1                               activin A receptor, type I             90   \n",
       "\n",
       "                   Genome Location  Tier Hallmark  Chr Band Somatic Germline  \\\n",
       "Gene Symbol                                                                    \n",
       "ABI1          10:26746593-26860935     1      Yes     12.10     yes      NaN   \n",
       "ABL1         9:130713946-130885683     1      Yes     34.12     yes      NaN   \n",
       "ABL2         1:179099327-179229601     1      NaN     25.20     yes      NaN   \n",
       "ACKR3        2:236569641-236582358     1      Yes     37.30     yes      NaN   \n",
       "ACVR1        2:157736444-157875111     1      Yes     24.10     yes      NaN   \n",
       "\n",
       "            Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome  \\\n",
       "Gene Symbol                                                                \n",
       "ABI1                          AML                    NaN             NaN   \n",
       "ABL1              CML, ALL, T-ALL                    NaN             NaN   \n",
       "ABL2                          AML                    NaN             NaN   \n",
       "ACKR3                      lipoma                    NaN             NaN   \n",
       "ACVR1                        DIPG                    NaN             NaN   \n",
       "\n",
       "            Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "Gene Symbol                                                                   \n",
       "ABI1                  L                Dom       TSG, fusion              T   \n",
       "ABL1                  L                Dom  oncogene, fusion         T, Mis   \n",
       "ABL2                  L                Dom  oncogene, fusion              T   \n",
       "ACKR3                 M                Dom  oncogene, fusion              T   \n",
       "ACVR1                 O                Dom          oncogene            Mis   \n",
       "\n",
       "            Translocation Partner Other Germline Mut  \\\n",
       "Gene Symbol                                            \n",
       "ABI1                        KMT2A                NaN   \n",
       "ABL1            BCR, ETV6, NUP214                NaN   \n",
       "ABL2                         ETV6                NaN   \n",
       "ACKR3                       HMGA2                NaN   \n",
       "ACVR1                         NaN                yes   \n",
       "\n",
       "                                    Other Syndrome  \\\n",
       "Gene Symbol                                          \n",
       "ABI1                                           NaN   \n",
       "ABL1                                           NaN   \n",
       "ABL2                                           NaN   \n",
       "ACKR3                                          NaN   \n",
       "ACVR1        Fibrodysplasia ossificans progressiva   \n",
       "\n",
       "                                                      Synonyms  \n",
       "Gene Symbol                                                     \n",
       "ABI1         10006,ABI-1,ABI1,E3B1,ENSG00000136754.17,Q8IZP...  \n",
       "ABL1         25,ABL,ABL1,ENSG00000097007.17,JTK7,P00519,c-A...  \n",
       "ABL2                27,ABL2,ABLL,ARG,ENSG00000143322.19,P42684  \n",
       "ACKR3        57007,ACKR3,CMKOR1,CXCR7,ENSG00000144476.5,GPR...  \n",
       "ACVR1        90,ACVR1,ACVR1A,ACVRLK2,ALK2,ENSG00000115170.1...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df = pd.read_csv(\n",
    "    cfg.cosmic_raw_file, sep='\\t', index_col=0\n",
    ")\n",
    "\n",
    "cosmic_df = cosmic_df[\n",
    "    # use only tier 1 genes\n",
    "    ((cosmic_df.Tier == 1) &\n",
    "    # drop genes without a catalogued somatic mutation\n",
    "     (cosmic_df.Somatic == 'yes') &\n",
    "    # drop genes that are only observed in cancer as fusions\n",
    "    # (we're not calling fusion genes in our mutation data)\n",
    "     (cosmic_df['Role in Cancer'] != 'fusion'))\n",
    "].copy()\n",
    "     \n",
    "print(cosmic_df.shape)\n",
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG, fusion' 'oncogene, fusion' 'oncogene' 'TSG' 'oncogene, TSG, fusion'\n",
      " 'oncogene, TSG']\n",
      "['TSG' 'oncogene' 'oncogene, TSG']\n"
     ]
    }
   ],
   "source": [
    "print(cosmic_df['Role in Cancer'].unique())\n",
    "\n",
    "# if a gene is annotated as an oncogene/TSG and a fusion gene, just\n",
    "# get rid of the fusion component\n",
    "# we'll resolve the dual annotated oncogene/TSG genes later\n",
    "cosmic_df['Role in Cancer'] = cosmic_df['Role in Cancer'].str.replace(', fusion', '')\n",
    "print(cosmic_df['Role in Cancer'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download Bailey et al. data\n",
    "\n",
    "This is a supplementary table from [the TCGA Pan-Cancer Atlas driver gene analysis](https://www.sciencedirect.com/science/article/pii/S009286741830237X). The table contains genes identified as cancer drivers by taking the consensus of existing driver identification methods, in addition to manual curation as described in the paper. The table also contains oncogene/TSG predictions for these genes, using the [20/20+ method](https://2020plus.readthedocs.io/en/latest/).\n",
    "\n",
    "This table (Excel file) was also directly downloaded from the paper's supplementary data, as Cell doesn't seem to provide a straightforward API (to my knowledge anyway)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(782, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jake/anaconda3/envs/mpmp/lib/python3.7/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>Note about previous publication</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEY</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1_PANCAN</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rescued</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Evidence from OncoImpact/DriverNET overlap (SN...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>0.053030</td>\n",
       "      <td>0.007490</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B_PANCAN</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>COADREAD</td>\n",
       "      <td>tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.028481</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 22810696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.031073</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in private communication about integrati...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene    Cancer classification  Decision  Tissue Frequency  \\\n",
       "KEY                                                                            \n",
       "ABL1_PANCAN        ABL1    PANCAN            NaN   rescued               NaN   \n",
       "ACVR1_UCEC        ACVR1      UCEC       oncogene  official          0.053030   \n",
       "ACVR1B_PANCAN    ACVR1B    PANCAN   possible tsg  official               NaN   \n",
       "ACVR2A_COADREAD  ACVR2A  COADREAD            tsg  official          0.028481   \n",
       "ACVR2A_LIHC      ACVR2A      LIHC   possible tsg  official          0.031073   \n",
       "\n",
       "                 Pancan Frequency  Consensus Score  Correlation adusted score  \\\n",
       "KEY                                                                             \n",
       "ABL1_PANCAN              0.011675              0.0                        NaN   \n",
       "ACVR1_UCEC               0.007490              1.5                        1.5   \n",
       "ACVR1B_PANCAN            0.010904              1.0                        0.0   \n",
       "ACVR2A_COADREAD          0.013988              1.5                        1.5   \n",
       "ACVR2A_LIHC              0.013988              1.5                        1.5   \n",
       "\n",
       "                 Novel                                       Rescue Notes  \\\n",
       "KEY                                                                         \n",
       "ABL1_PANCAN        0.0  Evidence from OncoImpact/DriverNET overlap (SN...   \n",
       "ACVR1_UCEC         0.0                                                NaN   \n",
       "ACVR1B_PANCAN      0.0                                                NaN   \n",
       "ACVR2A_COADREAD    0.0                                                NaN   \n",
       "ACVR2A_LIHC        0.0                                                NaN   \n",
       "\n",
       "                                   Note about previous publication  \n",
       "KEY                                                                 \n",
       "ABL1_PANCAN                                                    NaN  \n",
       "ACVR1_UCEC                                                       0  \n",
       "ACVR1B_PANCAN                                    Found in 24132290  \n",
       "ACVR2A_COADREAD                                  Found in 22810696  \n",
       "ACVR2A_LIHC      Found in private communication about integrati...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_df = pd.read_excel(\n",
    "    cfg.bailey_raw_file,\n",
    "    engine='openpyxl', sheet_name='Table S1', index_col='KEY', header=3\n",
    ")\n",
    "class_df.drop(\n",
    "    class_df.columns[class_df.columns.str.contains('Unnamed')],\n",
    "    axis=1, inplace=True\n",
    ")\n",
    "class_df.rename(columns={'Tumor suppressor or oncogene prediction (by 20/20+)':\n",
    "                         'classification'},\n",
    "                inplace=True)\n",
    "\n",
    "print(class_df.shape)\n",
    "class_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(170, 10)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>Note about previous publication</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1B</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.484499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010133</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMER1</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Cancer classification  Decision  Tissue Frequency  Pancan Frequency  \\\n",
       "Gene                                                                          \n",
       "ACVR1B  PANCAN            TSG  official               NaN          0.010904   \n",
       "ACVR2A  PANCAN            TSG  official               NaN          0.013988   \n",
       "AJUBA   PANCAN            TSG  official               NaN          0.009032   \n",
       "AKT1    PANCAN       Oncogene  official               NaN          0.010133   \n",
       "AMER1   PANCAN            TSG  official               NaN          0.027426   \n",
       "\n",
       "        Consensus Score  Correlation adusted score  Novel Rescue Notes  \\\n",
       "Gene                                                                     \n",
       "ACVR1B              1.0                   0.000000    0.0          NaN   \n",
       "ACVR2A              2.0                   2.000000    0.0          NaN   \n",
       "AJUBA               2.0                   1.484499    0.0          NaN   \n",
       "AKT1                2.5                   2.500000    0.0          NaN   \n",
       "AMER1               2.0                   2.000000    0.0          NaN   \n",
       "\n",
       "       Note about previous publication  \n",
       "Gene                                    \n",
       "ACVR1B               Found in 24132290  \n",
       "ACVR2A               Found in 24132290  \n",
       "AJUBA                Found in 24132290  \n",
       "AKT1                 Found in 24132290  \n",
       "AMER1                                0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bailey_df = (\n",
    "    class_df[((class_df.Cancer == 'PANCAN') &\n",
    "             (~class_df.classification.isna()))]\n",
    "      .set_index('Gene')\n",
    ").copy()\n",
    "\n",
    "# this is the best classification we have to go on for these genes, so if\n",
    "# a gene is labeled as \"possible X\", we'll just consider it X\n",
    "bailey_df['classification'] = (\n",
    "    bailey_df['classification'].str.replace('possible ', '')\n",
    "                                   .replace('tsg', 'TSG')\n",
    "                                   .replace('oncogene', 'Oncogene')\n",
    ")\n",
    "\n",
    "print(bailey_df.shape)\n",
    "bailey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas.api.types import is_datetime64_any_dtype\n",
    "\n",
    "# make sure no gene names were converted to dates, since excel does this sometimes\n",
    "assert not is_datetime64_any_dtype(bailey_df.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Vogelstein et al. data\n",
    "\n",
    "This data originally came from [Vogelstein et al. 2013](https://www.science.org/doi/10.1126/science.1235122), and is available as a tsv in [the `pancancer` repo](https://github.com/greenelab/pancancer/blob/master/data/vogelstein_cancergenes.tsv). Oncogene/TSG annotations are also 20/20+ predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(125, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th># Mutated Tumor Samples**</th>\n",
       "      <th>Ocogene score*</th>\n",
       "      <th>Tumor Suppressor Gene score*</th>\n",
       "      <th>classification</th>\n",
       "      <th>Core pathway</th>\n",
       "      <th>Process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>c-abl oncogene 1, receptor tyrosine kinase</td>\n",
       "      <td>851</td>\n",
       "      <td>0.926904</td>\n",
       "      <td>0.003046</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Cell Cycle/Apoptosis</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>activin A receptor, type IB</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.423077</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TGF-b</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>v-akt murine thymoma viral oncogene homolog 1</td>\n",
       "      <td>155</td>\n",
       "      <td>0.929487</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>anaplastic lymphoma receptor tyrosine kinase</td>\n",
       "      <td>189</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K; RAS</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>adenomatous polyposis coli</td>\n",
       "      <td>2561</td>\n",
       "      <td>0.024553</td>\n",
       "      <td>0.917222</td>\n",
       "      <td>TSG</td>\n",
       "      <td>APC</td>\n",
       "      <td>Cell Fate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene                                      Gene Name  \\\n",
       "0    ABL1     c-abl oncogene 1, receptor tyrosine kinase   \n",
       "1  ACVR1B                    activin A receptor, type IB   \n",
       "2    AKT1  v-akt murine thymoma viral oncogene homolog 1   \n",
       "3     ALK   anaplastic lymphoma receptor tyrosine kinase   \n",
       "4     APC                     adenomatous polyposis coli   \n",
       "\n",
       "   # Mutated Tumor Samples**  Ocogene score*   Tumor Suppressor Gene score*   \\\n",
       "0                        851         0.926904                       0.003046   \n",
       "1                         17         0.000000                       0.423077   \n",
       "2                        155         0.929487                       0.006410   \n",
       "3                        189         0.720000                       0.010000   \n",
       "4                       2561         0.024553                       0.917222   \n",
       "\n",
       "  classification          Core pathway        Process  \n",
       "0       Oncogene  Cell Cycle/Apoptosis  Cell Survival  \n",
       "1            TSG                TGF-b   Cell Survival  \n",
       "2       Oncogene                  PI3K  Cell Survival  \n",
       "3       Oncogene             PI3K; RAS  Cell Survival  \n",
       "4            TSG                   APC      Cell Fate  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mpmp.utilities.data_utilities as du\n",
    "\n",
    "vogelstein_df = du.load_vogelstein()\n",
    "\n",
    "print(vogelstein_df.shape)\n",
    "vogelstein_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean up oncogene/TSG annotations and merge datasets\n",
    "\n",
    "We want as many genes as possible to be annotated as _either_ an oncogene or TSG, so we know whether to use copy gain or copy loss data to define relevant CNV info. \n",
    "\n",
    "So, here, we will:\n",
    "\n",
    "1. Drop genes that are annotated only as fusion genes (since we're not calling fusions at this time)  \n",
    "2. Try to resolve genes that are annotated as both oncogene/TSG (usually context/cancer type specific) into their most likely pan-cancer category  \n",
    "3. For genes that can't be resolved confidently, we'll keep them as \"oncogene, TSG\" and run our scripts using both copy gain and copy loss downstream.\n",
    "4. Merge the three datasets together, ensuring that annotations are concordant (or resolving them if not)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Oncogene' 'TSG']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th># Mutated Tumor Samples**</th>\n",
       "      <th>Ocogene score*</th>\n",
       "      <th>Tumor Suppressor Gene score*</th>\n",
       "      <th>classification</th>\n",
       "      <th>Core pathway</th>\n",
       "      <th>Process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>c-abl oncogene 1, receptor tyrosine kinase</td>\n",
       "      <td>851</td>\n",
       "      <td>0.926904</td>\n",
       "      <td>0.003046</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Cell Cycle/Apoptosis</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>activin A receptor, type IB</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.423077</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TGF-b</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>v-akt murine thymoma viral oncogene homolog 1</td>\n",
       "      <td>155</td>\n",
       "      <td>0.929487</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>anaplastic lymphoma receptor tyrosine kinase</td>\n",
       "      <td>189</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K; RAS</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>adenomatous polyposis coli</td>\n",
       "      <td>2561</td>\n",
       "      <td>0.024553</td>\n",
       "      <td>0.917222</td>\n",
       "      <td>TSG</td>\n",
       "      <td>APC</td>\n",
       "      <td>Cell Fate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene                                      Gene Name  \\\n",
       "0    ABL1     c-abl oncogene 1, receptor tyrosine kinase   \n",
       "1  ACVR1B                    activin A receptor, type IB   \n",
       "2    AKT1  v-akt murine thymoma viral oncogene homolog 1   \n",
       "3     ALK   anaplastic lymphoma receptor tyrosine kinase   \n",
       "4     APC                     adenomatous polyposis coli   \n",
       "\n",
       "   # Mutated Tumor Samples**  Ocogene score*   Tumor Suppressor Gene score*   \\\n",
       "0                        851         0.926904                       0.003046   \n",
       "1                         17         0.000000                       0.423077   \n",
       "2                        155         0.929487                       0.006410   \n",
       "3                        189         0.720000                       0.010000   \n",
       "4                       2561         0.024553                       0.917222   \n",
       "\n",
       "  classification          Core pathway        Process  \n",
       "0       Oncogene  Cell Cycle/Apoptosis  Cell Survival  \n",
       "1            TSG                TGF-b   Cell Survival  \n",
       "2       Oncogene                  PI3K  Cell Survival  \n",
       "3       Oncogene             PI3K; RAS  Cell Survival  \n",
       "4            TSG                   APC      Cell Fate  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first merge bailey and vogelstein datasets\n",
    "# these are both \"confidently\" annotated (everything is either an oncogene or TSG),\n",
    "# so we don't need to worry about resolving ambiguous annotations\n",
    "print(vogelstein_df.classification.unique())\n",
    "vogelstein_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG' 'Oncogene']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>Note about previous publication</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1B</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.484499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010133</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Found in 24132290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMER1</th>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Cancer classification  Decision  Tissue Frequency  Pancan Frequency  \\\n",
       "Gene                                                                          \n",
       "ACVR1B  PANCAN            TSG  official               NaN          0.010904   \n",
       "ACVR2A  PANCAN            TSG  official               NaN          0.013988   \n",
       "AJUBA   PANCAN            TSG  official               NaN          0.009032   \n",
       "AKT1    PANCAN       Oncogene  official               NaN          0.010133   \n",
       "AMER1   PANCAN            TSG  official               NaN          0.027426   \n",
       "\n",
       "        Consensus Score  Correlation adusted score  Novel Rescue Notes  \\\n",
       "Gene                                                                     \n",
       "ACVR1B              1.0                   0.000000    0.0          NaN   \n",
       "ACVR2A              2.0                   2.000000    0.0          NaN   \n",
       "AJUBA               2.0                   1.484499    0.0          NaN   \n",
       "AKT1                2.5                   2.500000    0.0          NaN   \n",
       "AMER1               2.0                   2.000000    0.0          NaN   \n",
       "\n",
       "       Note about previous publication  \n",
       "Gene                                    \n",
       "ACVR1B               Found in 24132290  \n",
       "ACVR2A               Found in 24132290  \n",
       "AJUBA                Found in 24132290  \n",
       "AKT1                 Found in 24132290  \n",
       "AMER1                                0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(bailey_df.classification.unique())\n",
    "bailey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(218, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vogelstein_classification</th>\n",
       "      <th>bailey_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B</th>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       vogelstein_classification bailey_classification\n",
       "ABL1                    Oncogene                   NaN\n",
       "ACVR1B                       TSG                   TSG\n",
       "ACVR2A                       NaN                   TSG\n",
       "AJUBA                        NaN                   TSG\n",
       "AKT1                    Oncogene              Oncogene"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first merge dataframes, then resolve classifications\n",
    "vogelstein_bailey_df = (\n",
    "    vogelstein_df.loc[:, ['gene', 'classification']]\n",
    "      .set_index('gene')\n",
    "      .merge(bailey_df.loc[:, ['classification']],\n",
    "             how='outer', left_index=True, right_index=True)\n",
    "      .rename(columns={'classification_x': 'vogelstein_classification',\n",
    "                       'classification_y': 'bailey_classification'})\n",
    ")\n",
    "\n",
    "print(vogelstein_bailey_df.shape)\n",
    "vogelstein_bailey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Oncogene' 'TSG' 'check']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vogelstein_classification</th>\n",
       "      <th>bailey_classification</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B</th>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       vogelstein_classification bailey_classification classification\n",
       "ABL1                    Oncogene                   NaN       Oncogene\n",
       "ACVR1B                       TSG                   TSG            TSG\n",
       "ACVR2A                       NaN                   TSG            TSG\n",
       "AJUBA                        NaN                   TSG            TSG\n",
       "AKT1                    Oncogene              Oncogene       Oncogene"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def merge_classifications(row):\n",
    "    if row['vogelstein_classification'] == row['bailey_classification']:\n",
    "        return row['vogelstein_classification']\n",
    "    elif pd.isna(row['vogelstein_classification']):\n",
    "        return row['bailey_classification']\n",
    "    elif pd.isna(row['bailey_classification']):\n",
    "        return row['vogelstein_classification']\n",
    "    elif row['vogelstein_classification'] != row['bailey_classification']:\n",
    "        # if the datasets disagree, we can resolve these manually\n",
    "        # or just run them as \"oncogene, TSG\"\n",
    "        return 'check'\n",
    "    else:\n",
    "        # not sure how this would happen\n",
    "        return pd.NA\n",
    "    \n",
    "vogelstein_bailey_df['classification'] = (\n",
    "    vogelstein_bailey_df.apply(merge_classifications, axis='columns')\n",
    ")\n",
    "\n",
    "print(vogelstein_bailey_df.classification.unique())\n",
    "vogelstein_bailey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vogelstein_classification</th>\n",
       "      <th>bailey_classification</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DNMT3A</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>TSG</td>\n",
       "      <td>check</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>TSG</td>\n",
       "      <td>check</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMARCA4</th>\n",
       "      <td>TSG</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>check</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WT1</th>\n",
       "      <td>TSG</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>check</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        vogelstein_classification bailey_classification classification\n",
       "DNMT3A                   Oncogene                   TSG          check\n",
       "JAK1                     Oncogene                   TSG          check\n",
       "SMARCA4                       TSG              Oncogene          check\n",
       "WT1                           TSG              Oncogene          check"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# examples where the datasets disagree, there shouldn't be too many of them\n",
    "print(vogelstein_bailey_df[vogelstein_bailey_df.classification == 'check'].shape)\n",
    "vogelstein_bailey_df[vogelstein_bailey_df.classification == 'check']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Oncogene' 'TSG' 'Oncogene, TSG']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B</th>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A</th>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA</th>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       classification\n",
       "ABL1         Oncogene\n",
       "ACVR1B            TSG\n",
       "ACVR2A            TSG\n",
       "AJUBA             TSG\n",
       "AKT1         Oncogene"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# COSMIC CGC classifies DNMT3A as a TSG at the pan-cancer level,\n",
    "# so we'll go with that\n",
    "# https://cancer.sanger.ac.uk/cosmic/census-page/DNMT3A\n",
    "vogelstein_bailey_df.loc['DNMT3A', 'classification'] = 'TSG'\n",
    "\n",
    "# COSMIC CGC says JAK1 can act as either a TSG or an oncogene, so\n",
    "# we'll run it as both\n",
    "# https://cancer.sanger.ac.uk/cosmic/census-page/JAK1\n",
    "vogelstein_bailey_df.loc['JAK1', 'classification'] = 'Oncogene, TSG'\n",
    "\n",
    "# SMARCA4 has been characterized as both a tumor suppressor and an\n",
    "# oncogene, depending on context\n",
    "# https://www.nature.com/articles/s41388-021-01875-6\n",
    "# https://www.frontiersin.org/articles/10.3389/fimmu.2021.762598/full\n",
    "vogelstein_bailey_df.loc['SMARCA4', 'classification'] = 'Oncogene, TSG'\n",
    "\n",
    "# WT1 can also act as either a TSG or an oncogene depending on the\n",
    "# cancer type/context:\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4352850/\n",
    "# or even within the same cancer type:\n",
    "# https://www.nature.com/articles/2404624\n",
    "vogelstein_bailey_df.loc['WT1', 'classification'] = 'Oncogene, TSG'\n",
    "\n",
    "\n",
    "vogelstein_bailey_df.drop(\n",
    "    columns=['vogelstein_classification', 'bailey_classification'],\n",
    "    inplace=True\n",
    ")\n",
    "print(vogelstein_bailey_df.classification.unique())\n",
    "vogelstein_bailey_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(511, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vb_classification</th>\n",
       "      <th>cosmic_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      vb_classification cosmic_classification\n",
       "ABI1                NaN                   TSG\n",
       "ABL1           Oncogene              Oncogene\n",
       "ABL2                NaN              Oncogene\n",
       "ACKR3               NaN              Oncogene\n",
       "ACVR1               NaN              Oncogene"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# now merge with cosmic CGC genes\n",
    "merged_df = (vogelstein_bailey_df\n",
    "  .merge(cosmic_df.loc[:, ['Role in Cancer']],\n",
    "         how='outer', left_index=True, right_index=True)\n",
    "  .rename(columns={'classification': 'vb_classification',\n",
    "                   'Role in Cancer': 'cosmic_classification'})\n",
    ")\n",
    "merged_df['cosmic_classification'] = (\n",
    "    merged_df['cosmic_classification'].str.replace('oncogene', 'Oncogene')\n",
    ")\n",
    "\n",
    "print(merged_df.shape)\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG' 'Oncogene' 'Oncogene, TSG']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>vb_classification</th>\n",
       "      <th>cosmic_classification</th>\n",
       "      <th>classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Oncogene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      vb_classification cosmic_classification classification\n",
       "ABI1                NaN                   TSG            TSG\n",
       "ABL1           Oncogene              Oncogene       Oncogene\n",
       "ABL2                NaN              Oncogene       Oncogene\n",
       "ACKR3               NaN              Oncogene       Oncogene\n",
       "ACVR1               NaN              Oncogene       Oncogene"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def merge_all(row):\n",
    "    if row['vb_classification'] == row['cosmic_classification']:\n",
    "        return row['vb_classification']\n",
    "    elif pd.isna(row['vb_classification']):\n",
    "        return row['cosmic_classification']\n",
    "    elif pd.isna(row['cosmic_classification']):\n",
    "        return row['vb_classification']\n",
    "    elif row['cosmic_classification'] == 'Oncogene, TSG':\n",
    "        # for ambiguous cosmic examples, just go with the\n",
    "        # bailey/vogelstein annotation\n",
    "        return row['vb_classification']\n",
    "    elif row['vb_classification'] != row['cosmic_classification']:\n",
    "        # if the datasets disagree, go with the cosmic annotation\n",
    "        # these are typically manually curated and likely to be more\n",
    "        # generally applicable\n",
    "        return row['cosmic_classification']\n",
    "    else:\n",
    "        # not sure how this would happen\n",
    "        return pd.NA\n",
    "    \n",
    "merged_df['classification'] = (\n",
    "    merged_df.apply(merge_all, axis='columns')\n",
    ")\n",
    "\n",
    "print(merged_df.classification.unique())\n",
    "merged_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "(merged_df\n",
    "  .loc[:, 'classification']\n",
    "  .rename_axis('gene')\n",
    "  .to_csv(cfg.merged_cancer_genes, sep='\\t')\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mpmp]",
   "language": "python",
   "name": "conda-env-mpmp-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
