{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download cancer gene sets\n",
    "\n",
    "We want to download the set of cancer-associated genes from the [COSMIC Cancer Gene Census](https://cancer.sanger.ac.uk/cosmic/census), and from [Bailey et al. 2018](https://www.sciencedirect.com/science/article/pii/S009286741830237X), in order to use these genes in our experiments as a comparison/complement to the Vogelstein et al. gene set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import mpmp.config as cfg\n",
    "import mpmp.utilities.data_utilities as du\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download COSMIC CGC data\n",
    "\n",
    "We downloaded the original CGC data directly from the Sanger Institute website linked above - you need to create an account there to download the .tsv file, so we can't do it programmatically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(445, 19)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Entrez GeneId</th>\n",
       "      <th>Genome Location</th>\n",
       "      <th>Tier</th>\n",
       "      <th>Hallmark</th>\n",
       "      <th>Chr Band</th>\n",
       "      <th>Somatic</th>\n",
       "      <th>Germline</th>\n",
       "      <th>Tumour Types(Somatic)</th>\n",
       "      <th>Tumour Types(Germline)</th>\n",
       "      <th>Cancer Syndrome</th>\n",
       "      <th>Tissue Type</th>\n",
       "      <th>Molecular Genetics</th>\n",
       "      <th>Role in Cancer</th>\n",
       "      <th>Mutation Types</th>\n",
       "      <th>Translocation Partner</th>\n",
       "      <th>Other Germline Mut</th>\n",
       "      <th>Other Syndrome</th>\n",
       "      <th>Synonyms</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABI1</th>\n",
       "      <td>abl-interactor 1</td>\n",
       "      <td>10006</td>\n",
       "      <td>10:26746593-26860935</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>12.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>TSG, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>KMT2A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10006,ABI-1,ABI1,E3B1,ENSG00000136754.17,Q8IZP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene h...</td>\n",
       "      <td>25</td>\n",
       "      <td>9:130713946-130885683</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>34.12</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CML, ALL, T-ALL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T, Mis</td>\n",
       "      <td>BCR, ETV6, NUP214</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25,ABL,ABL1,ENSG00000097007.17,JTK7,P00519,c-A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL2</th>\n",
       "      <td>c-abl oncogene 2, non-receptor tyrosine kinase</td>\n",
       "      <td>27</td>\n",
       "      <td>1:179099327-179229601</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.20</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AML</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>L</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>ETV6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27,ABL2,ABLL,ARG,ENSG00000143322.19,P42684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACKR3</th>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>57007</td>\n",
       "      <td>2:236569641-236582358</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>37.30</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lipoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene, fusion</td>\n",
       "      <td>T</td>\n",
       "      <td>HMGA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57007,ACKR3,CMKOR1,CXCR7,ENSG00000144476.5,GPR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1</th>\n",
       "      <td>activin A receptor, type I</td>\n",
       "      <td>90</td>\n",
       "      <td>2:157736444-157875111</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>24.10</td>\n",
       "      <td>yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DIPG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O</td>\n",
       "      <td>Dom</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>Mis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yes</td>\n",
       "      <td>Fibrodysplasia ossificans progressiva</td>\n",
       "      <td>90,ACVR1,ACVR1A,ACVRLK2,ALK2,ENSG00000115170.1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          Name  Entrez GeneId  \\\n",
       "Gene Symbol                                                                     \n",
       "ABI1                                          abl-interactor 1          10006   \n",
       "ABL1         v-abl Abelson murine leukemia viral oncogene h...             25   \n",
       "ABL2            c-abl oncogene 2, non-receptor tyrosine kinase             27   \n",
       "ACKR3                            atypical chemokine receptor 3          57007   \n",
       "ACVR1                               activin A receptor, type I             90   \n",
       "\n",
       "                   Genome Location  Tier Hallmark  Chr Band Somatic Germline  \\\n",
       "Gene Symbol                                                                    \n",
       "ABI1          10:26746593-26860935     1      Yes     12.10     yes      NaN   \n",
       "ABL1         9:130713946-130885683     1      Yes     34.12     yes      NaN   \n",
       "ABL2         1:179099327-179229601     1      NaN     25.20     yes      NaN   \n",
       "ACKR3        2:236569641-236582358     1      Yes     37.30     yes      NaN   \n",
       "ACVR1        2:157736444-157875111     1      Yes     24.10     yes      NaN   \n",
       "\n",
       "            Tumour Types(Somatic) Tumour Types(Germline) Cancer Syndrome  \\\n",
       "Gene Symbol                                                                \n",
       "ABI1                          AML                    NaN             NaN   \n",
       "ABL1              CML, ALL, T-ALL                    NaN             NaN   \n",
       "ABL2                          AML                    NaN             NaN   \n",
       "ACKR3                      lipoma                    NaN             NaN   \n",
       "ACVR1                        DIPG                    NaN             NaN   \n",
       "\n",
       "            Tissue Type Molecular Genetics    Role in Cancer Mutation Types  \\\n",
       "Gene Symbol                                                                   \n",
       "ABI1                  L                Dom       TSG, fusion              T   \n",
       "ABL1                  L                Dom  oncogene, fusion         T, Mis   \n",
       "ABL2                  L                Dom  oncogene, fusion              T   \n",
       "ACKR3                 M                Dom  oncogene, fusion              T   \n",
       "ACVR1                 O                Dom          oncogene            Mis   \n",
       "\n",
       "            Translocation Partner Other Germline Mut  \\\n",
       "Gene Symbol                                            \n",
       "ABI1                        KMT2A                NaN   \n",
       "ABL1            BCR, ETV6, NUP214                NaN   \n",
       "ABL2                         ETV6                NaN   \n",
       "ACKR3                       HMGA2                NaN   \n",
       "ACVR1                         NaN                yes   \n",
       "\n",
       "                                    Other Syndrome  \\\n",
       "Gene Symbol                                          \n",
       "ABI1                                           NaN   \n",
       "ABL1                                           NaN   \n",
       "ABL2                                           NaN   \n",
       "ACKR3                                          NaN   \n",
       "ACVR1        Fibrodysplasia ossificans progressiva   \n",
       "\n",
       "                                                      Synonyms  \n",
       "Gene Symbol                                                     \n",
       "ABI1         10006,ABI-1,ABI1,E3B1,ENSG00000136754.17,Q8IZP...  \n",
       "ABL1         25,ABL,ABL1,ENSG00000097007.17,JTK7,P00519,c-A...  \n",
       "ABL2                27,ABL2,ABLL,ARG,ENSG00000143322.19,P42684  \n",
       "ACKR3        57007,ACKR3,CMKOR1,CXCR7,ENSG00000144476.5,GPR...  \n",
       "ACVR1        90,ACVR1,ACVR1A,ACVRLK2,ALK2,ENSG00000115170.1...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cosmic_df = pd.read_csv(\n",
    "    cfg.cosmic_genes_file, sep='\\t', index_col=0\n",
    ")\n",
    "\n",
    "cosmic_df = cosmic_df[\n",
    "    # use only tier 1 genes\n",
    "    ((cosmic_df.Tier == 1) &\n",
    "    # drop genes without a catalogued somatic mutation\n",
    "     (cosmic_df.Somatic == 'yes') &\n",
    "    # drop genes that are only observed in cancer as fusions\n",
    "    # (we're not calling fusion genes in our mutation data)\n",
    "     (cosmic_df['Role in Cancer'] != 'fusion'))\n",
    "].copy()\n",
    "     \n",
    "print(cosmic_df.shape)\n",
    "cosmic_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TSG, fusion' 'oncogene, fusion' 'oncogene' 'TSG' 'oncogene, TSG, fusion'\n",
      " 'oncogene, TSG']\n",
      "['TSG' 'oncogene' 'oncogene, TSG']\n"
     ]
    }
   ],
   "source": [
    "print(cosmic_df['Role in Cancer'].unique())\n",
    "\n",
    "# if a gene is annotated as an oncogene/TSG and a fusion gene, just\n",
    "# get rid of the fusion component\n",
    "# we'll resolve the dual annotated oncogene/TSG genes later\n",
    "cosmic_df['Role in Cancer'] = cosmic_df['Role in Cancer'].str.replace(', fusion', '')\n",
    "print(cosmic_df['Role in Cancer'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download Bailey et al. data\n",
    "\n",
    "This is a supplementary table from [the TCGA Pan-Cancer Atlas driver gene analysis](https://www.sciencedirect.com/science/article/pii/S009286741830237X). The table contains genes identified as cancer drivers by taking the consensus of existing driver identification methods, in addition to manual curation as described in the paper. The table also contains oncogene/TSG predictions for these genes, using the [20/20+ method](https://2020plus.readthedocs.io/en/latest/).\n",
    "\n",
    "This table (Excel file) was also directly downloaded from the paper's supplementary data, as Cell doesn't seem to provide a straightforward API (that I'm able to find)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(782, 25)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jake/anaconda3/envs/mpmp/lib/python3.7/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "      <th>Unnamed: 20</th>\n",
       "      <th>Unnamed: 21</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEY</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABL1_PANCAN</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rescued</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.011675</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Evidence from OncoImpact/DriverNET overlap (SN...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1_UCEC</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>UCEC</td>\n",
       "      <td>oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>0.053030</td>\n",
       "      <td>0.007490</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR1B_PANCAN</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_COADREAD</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>COADREAD</td>\n",
       "      <td>tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.028481</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_LIHC</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>LIHC</td>\n",
       "      <td>possible tsg</td>\n",
       "      <td>official</td>\n",
       "      <td>0.031073</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>1.5</td>\n",
       "      <td>1.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene    Cancer classification  Decision  Tissue Frequency  \\\n",
       "KEY                                                                            \n",
       "ABL1_PANCAN        ABL1    PANCAN            NaN   rescued               NaN   \n",
       "ACVR1_UCEC        ACVR1      UCEC       oncogene  official          0.053030   \n",
       "ACVR1B_PANCAN    ACVR1B    PANCAN   possible tsg  official               NaN   \n",
       "ACVR2A_COADREAD  ACVR2A  COADREAD            tsg  official          0.028481   \n",
       "ACVR2A_LIHC      ACVR2A      LIHC   possible tsg  official          0.031073   \n",
       "\n",
       "                 Pancan Frequency  Consensus Score  Correlation adusted score  \\\n",
       "KEY                                                                             \n",
       "ABL1_PANCAN              0.011675              0.0                        NaN   \n",
       "ACVR1_UCEC               0.007490              1.5                        1.5   \n",
       "ACVR1B_PANCAN            0.010904              1.0                        0.0   \n",
       "ACVR2A_COADREAD          0.013988              1.5                        1.5   \n",
       "ACVR2A_LIHC              0.013988              1.5                        1.5   \n",
       "\n",
       "                 Novel                                       Rescue Notes  \\\n",
       "KEY                                                                         \n",
       "ABL1_PANCAN        0.0  Evidence from OncoImpact/DriverNET overlap (SN...   \n",
       "ACVR1_UCEC         0.0                                                NaN   \n",
       "ACVR1B_PANCAN      0.0                                                NaN   \n",
       "ACVR2A_COADREAD    0.0                                                NaN   \n",
       "ACVR2A_LIHC        0.0                                                NaN   \n",
       "\n",
       "                 ... Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \\\n",
       "KEY              ...                                                      \n",
       "ABL1_PANCAN      ...         NaN          NaN          NaN          NaN   \n",
       "ACVR1_UCEC       ...         NaN          NaN          NaN          NaN   \n",
       "ACVR1B_PANCAN    ...         NaN          NaN          NaN          NaN   \n",
       "ACVR2A_COADREAD  ...         NaN          NaN          NaN          NaN   \n",
       "ACVR2A_LIHC      ...         NaN          NaN          NaN          NaN   \n",
       "\n",
       "                 Unnamed: 20  Unnamed: 21  Unnamed: 22  Unnamed: 23  \\\n",
       "KEY                                                                   \n",
       "ABL1_PANCAN              NaN          NaN          NaN          NaN   \n",
       "ACVR1_UCEC               NaN          NaN          NaN          NaN   \n",
       "ACVR1B_PANCAN            NaN          NaN          NaN          NaN   \n",
       "ACVR2A_COADREAD          NaN          NaN          NaN          NaN   \n",
       "ACVR2A_LIHC              NaN          NaN          NaN          NaN   \n",
       "\n",
       "                 Unnamed: 24  Unnamed: 25  \n",
       "KEY                                        \n",
       "ABL1_PANCAN              NaN          NaN  \n",
       "ACVR1_UCEC               NaN          NaN  \n",
       "ACVR1B_PANCAN            NaN          NaN  \n",
       "ACVR2A_COADREAD          NaN          NaN  \n",
       "ACVR2A_LIHC              NaN          NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class_df = pd.read_excel(\n",
    "    cfg.data_dir / '1-s2.0-S009286741830237X-mmc1.xlsx', \n",
    "    engine='openpyxl', sheet_name='Table S1', index_col='KEY', header=3\n",
    ")\n",
    "class_df.rename(columns={'Tumor suppressor or oncogene prediction (by 20/20+)':\n",
    "                         'classification'},\n",
    "                inplace=True)\n",
    "\n",
    "print(class_df.shape)\n",
    "class_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(170, 25)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Cancer</th>\n",
       "      <th>classification</th>\n",
       "      <th>Decision</th>\n",
       "      <th>Tissue Frequency</th>\n",
       "      <th>Pancan Frequency</th>\n",
       "      <th>Consensus Score</th>\n",
       "      <th>Correlation adusted score</th>\n",
       "      <th>Novel</th>\n",
       "      <th>Rescue Notes</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 16</th>\n",
       "      <th>Unnamed: 17</th>\n",
       "      <th>Unnamed: 18</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "      <th>Unnamed: 20</th>\n",
       "      <th>Unnamed: 21</th>\n",
       "      <th>Unnamed: 22</th>\n",
       "      <th>Unnamed: 23</th>\n",
       "      <th>Unnamed: 24</th>\n",
       "      <th>Unnamed: 25</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEY</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACVR1B_PANCAN</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010904</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACVR2A_PANCAN</th>\n",
       "      <td>ACVR2A</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.013988</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AJUBA_PANCAN</th>\n",
       "      <td>AJUBA</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.009032</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.484499</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1_PANCAN</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010133</td>\n",
       "      <td>2.5</td>\n",
       "      <td>2.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AMER1_PANCAN</th>\n",
       "      <td>AMER1</td>\n",
       "      <td>PANCAN</td>\n",
       "      <td>TSG</td>\n",
       "      <td>official</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027426</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Gene  Cancer classification  Decision  Tissue Frequency  \\\n",
       "KEY                                                                        \n",
       "ACVR1B_PANCAN  ACVR1B  PANCAN            TSG  official               NaN   \n",
       "ACVR2A_PANCAN  ACVR2A  PANCAN            TSG  official               NaN   \n",
       "AJUBA_PANCAN    AJUBA  PANCAN            TSG  official               NaN   \n",
       "AKT1_PANCAN      AKT1  PANCAN       Oncogene  official               NaN   \n",
       "AMER1_PANCAN    AMER1  PANCAN            TSG  official               NaN   \n",
       "\n",
       "               Pancan Frequency  Consensus Score  Correlation adusted score  \\\n",
       "KEY                                                                           \n",
       "ACVR1B_PANCAN          0.010904              1.0                   0.000000   \n",
       "ACVR2A_PANCAN          0.013988              2.0                   2.000000   \n",
       "AJUBA_PANCAN           0.009032              2.0                   1.484499   \n",
       "AKT1_PANCAN            0.010133              2.5                   2.500000   \n",
       "AMER1_PANCAN           0.027426              2.0                   2.000000   \n",
       "\n",
       "               Novel Rescue Notes  ... Unnamed: 16  Unnamed: 17  Unnamed: 18  \\\n",
       "KEY                                ...                                         \n",
       "ACVR1B_PANCAN    0.0          NaN  ...         NaN          NaN          NaN   \n",
       "ACVR2A_PANCAN    0.0          NaN  ...         NaN          NaN          NaN   \n",
       "AJUBA_PANCAN     0.0          NaN  ...         NaN          NaN          NaN   \n",
       "AKT1_PANCAN      0.0          NaN  ...         NaN          NaN          NaN   \n",
       "AMER1_PANCAN     0.0          NaN  ...         NaN          NaN          NaN   \n",
       "\n",
       "               Unnamed: 19  Unnamed: 20  Unnamed: 21  Unnamed: 22  \\\n",
       "KEY                                                                 \n",
       "ACVR1B_PANCAN          NaN          NaN          NaN          NaN   \n",
       "ACVR2A_PANCAN          NaN          NaN          NaN          NaN   \n",
       "AJUBA_PANCAN           NaN          NaN          NaN          NaN   \n",
       "AKT1_PANCAN            NaN          NaN          NaN          NaN   \n",
       "AMER1_PANCAN           NaN          NaN          NaN          NaN   \n",
       "\n",
       "               Unnamed: 23  Unnamed: 24  Unnamed: 25  \n",
       "KEY                                                   \n",
       "ACVR1B_PANCAN          NaN          NaN          NaN  \n",
       "ACVR2A_PANCAN          NaN          NaN          NaN  \n",
       "AJUBA_PANCAN           NaN          NaN          NaN  \n",
       "AKT1_PANCAN            NaN          NaN          NaN  \n",
       "AMER1_PANCAN           NaN          NaN          NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bailey_df = (\n",
    "    class_df[((class_df.Cancer == 'PANCAN') &\n",
    "             (~class_df.classification.isna()))]\n",
    ").copy()\n",
    "\n",
    "# this is the best classification we have to go on for these genes, so if\n",
    "# a gene is labeled as \"possible X\", we'll just consider it X\n",
    "bailey_df['classification'] = (\n",
    "    bailey_df['classification'].str.replace('possible ', '')\n",
    "                                   .replace('tsg', 'TSG')\n",
    "                                   .replace('oncogene', 'Oncogene')\n",
    ")\n",
    "\n",
    "print(bailey_df.shape)\n",
    "bailey_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load Vogelstein et al. data\n",
    "\n",
    "This data originally came from [Vogelstein et al. 2013](https://www.science.org/doi/10.1126/science.1235122). Oncogene/TSG annotations also come from 20/20+ predictions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(125, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th># Mutated Tumor Samples**</th>\n",
       "      <th>Ocogene score*</th>\n",
       "      <th>Tumor Suppressor Gene score*</th>\n",
       "      <th>classification</th>\n",
       "      <th>Core pathway</th>\n",
       "      <th>Process</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>c-abl oncogene 1, receptor tyrosine kinase</td>\n",
       "      <td>851</td>\n",
       "      <td>0.926904</td>\n",
       "      <td>0.003046</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>Cell Cycle/Apoptosis</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACVR1B</td>\n",
       "      <td>activin A receptor, type IB</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.423077</td>\n",
       "      <td>TSG</td>\n",
       "      <td>TGF-b</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AKT1</td>\n",
       "      <td>v-akt murine thymoma viral oncogene homolog 1</td>\n",
       "      <td>155</td>\n",
       "      <td>0.929487</td>\n",
       "      <td>0.006410</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALK</td>\n",
       "      <td>anaplastic lymphoma receptor tyrosine kinase</td>\n",
       "      <td>189</td>\n",
       "      <td>0.720000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>Oncogene</td>\n",
       "      <td>PI3K; RAS</td>\n",
       "      <td>Cell Survival</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>APC</td>\n",
       "      <td>adenomatous polyposis coli</td>\n",
       "      <td>2561</td>\n",
       "      <td>0.024553</td>\n",
       "      <td>0.917222</td>\n",
       "      <td>TSG</td>\n",
       "      <td>APC</td>\n",
       "      <td>Cell Fate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene                                      Gene Name  \\\n",
       "0    ABL1     c-abl oncogene 1, receptor tyrosine kinase   \n",
       "1  ACVR1B                    activin A receptor, type IB   \n",
       "2    AKT1  v-akt murine thymoma viral oncogene homolog 1   \n",
       "3     ALK   anaplastic lymphoma receptor tyrosine kinase   \n",
       "4     APC                     adenomatous polyposis coli   \n",
       "\n",
       "   # Mutated Tumor Samples**  Ocogene score*   Tumor Suppressor Gene score*   \\\n",
       "0                        851         0.926904                       0.003046   \n",
       "1                         17         0.000000                       0.423077   \n",
       "2                        155         0.929487                       0.006410   \n",
       "3                        189         0.720000                       0.010000   \n",
       "4                       2561         0.024553                       0.917222   \n",
       "\n",
       "  classification          Core pathway        Process  \n",
       "0       Oncogene  Cell Cycle/Apoptosis  Cell Survival  \n",
       "1            TSG                TGF-b   Cell Survival  \n",
       "2       Oncogene                  PI3K  Cell Survival  \n",
       "3       Oncogene             PI3K; RAS  Cell Survival  \n",
       "4            TSG                   APC      Cell Fate  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import mpmp.utilities.data_utilities as du\n",
    "\n",
    "vogelstein_df = du.load_vogelstein()\n",
    "\n",
    "print(vogelstein_df.shape)\n",
    "vogelstein_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mpmp]",
   "language": "python",
   "name": "conda-env-mpmp-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
