{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocess pan-cancer data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the downloaded data, update gene identifiers to Entrez, and curate sample IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "import mpmp.config as cfg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read TCGA Barcode Curation Information\n",
    "\n",
    "Extract information from TCGA barcodes - `cancer-type` and `sample-type`. See https://github.com/cognoma/cancer-data for more details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS Code</th>\n",
       "      <th>Source Site</th>\n",
       "      <th>Study Name</th>\n",
       "      <th>BCR</th>\n",
       "      <th>acronym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>01</td>\n",
       "      <td>International Genomics Consortium</td>\n",
       "      <td>ovarian serous cystadenocarcinoma</td>\n",
       "      <td>IGC</td>\n",
       "      <td>OV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>02</td>\n",
       "      <td>MD Anderson Cancer Center</td>\n",
       "      <td>glioblastoma multiforme</td>\n",
       "      <td>IGC</td>\n",
       "      <td>GBM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TSS Code                        Source Site  \\\n",
       "0       01  International Genomics Consortium   \n",
       "1       02          MD Anderson Cancer Center   \n",
       "\n",
       "                          Study Name  BCR acronym  \n",
       "0  ovarian serous cystadenocarcinoma  IGC      OV  \n",
       "1            glioblastoma multiforme  IGC     GBM  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Commit from https://github.com/cognoma/cancer-data/\n",
    "sample_commit = 'da832c5edc1ca4d3f665b038d15b19fced724f4c'\n",
    "url = 'https://raw.githubusercontent.com/cognoma/cancer-data/{}/mapping/tcga_cancertype_codes.csv'.format(sample_commit)\n",
    "cancer_types_df = pd.read_csv(url,\n",
    "                              dtype='str',\n",
    "                              keep_default_na=False)\n",
    "\n",
    "cancertype_codes_dict = dict(zip(cancer_types_df['TSS Code'],\n",
    "                                 cancer_types_df.acronym))\n",
    "cancer_types_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Code</th>\n",
       "      <th>Definition</th>\n",
       "      <th>Short Letter Code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>01</td>\n",
       "      <td>Primary Solid Tumor</td>\n",
       "      <td>TP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>02</td>\n",
       "      <td>Recurrent Solid Tumor</td>\n",
       "      <td>TR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Code             Definition Short Letter Code\n",
       "0   01    Primary Solid Tumor                TP\n",
       "1   02  Recurrent Solid Tumor                TR"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://raw.githubusercontent.com/cognoma/cancer-data/{}/mapping/tcga_sampletype_codes.csv'.format(sample_commit)\n",
    "sample_types_df = pd.read_csv(url, dtype='str')\n",
    "\n",
    "sampletype_codes_dict = dict(zip(sample_types_df.Code,\n",
    "                                 sample_types_df.Definition))\n",
    "sample_types_df.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read Entrez ID Curation Information\n",
    "\n",
    "Load curated gene names from versioned resource. See https://github.com/cognoma/genes for more details"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Commit from https://github.com/cognoma/genes\n",
    "genes_commit = 'ad9631bb4e77e2cdc5413b0d77cb8f7e93fc5bee'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20395, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>entrez_gene_id</th>\n",
       "      <th>symbol</th>\n",
       "      <th>description</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>synonyms</th>\n",
       "      <th>aliases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>19</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>12</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   entrez_gene_id symbol             description chromosome       gene_type  \\\n",
       "0               1   A1BG  alpha-1-B glycoprotein         19  protein-coding   \n",
       "1               2    A2M   alpha-2-macroglobulin         12  protein-coding   \n",
       "\n",
       "                    synonyms  \\\n",
       "0       A1B|ABG|GAB|HYST2477   \n",
       "1  A2MD|CPAMD5|FWP007|S863-7   \n",
       "\n",
       "                                             aliases  \n",
       "0  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...  \n",
       "1  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = 'https://raw.githubusercontent.com/cognoma/genes/{}/data/genes.tsv'.format(genes_commit)\n",
    "gene_df = pd.read_csv(url, sep='\\t')\n",
    "\n",
    "# Only consider protein-coding genes\n",
    "gene_df = (\n",
    "    gene_df.query(\"gene_type == 'protein-coding'\")\n",
    ")\n",
    "\n",
    "print(gene_df.shape)\n",
    "gene_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene updater - define up to date Entrez gene identifiers where appropriate\n",
    "url = 'https://raw.githubusercontent.com/cognoma/genes/{}/data/updater.tsv'.format(genes_commit)\n",
    "updater_df = pd.read_csv(url, sep='\\t')\n",
    "\n",
    "old_to_new_entrez = dict(zip(updater_df.old_entrez_gene_id,\n",
    "                             updater_df.new_entrez_gene_id))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and process gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>size</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>mirna_sample</th>\n",
       "      <td>55d9bf6f-0712-4315-b588-e6f8e295018e</td>\n",
       "      <td>PanCanAtlas_miRNA_sample_information_list.txt</td>\n",
       "      <td>02bb56712be34bcd58c50d90387aebde</td>\n",
       "      <td>553408</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                id  \\\n",
       "name                                                 \n",
       "mirna_sample  55d9bf6f-0712-4315-b588-e6f8e295018e   \n",
       "\n",
       "                                                   filename  \\\n",
       "name                                                          \n",
       "mirna_sample  PanCanAtlas_miRNA_sample_information_list.txt   \n",
       "\n",
       "                                           md5    size  \n",
       "name                                                    \n",
       "mirna_sample  02bb56712be34bcd58c50d90387aebde  553408  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first load manifest file, this tells us the filenames of the raw data files\n",
    "manifest_df = pd.read_csv(os.path.join(cfg.data_dir, 'manifest.tsv'),\n",
    "                          sep='\\t', index_col=0)\n",
    "manifest_df.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(20531, 11069)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-OR-A5J1-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J2-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J3-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J5-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J6-01A-31R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J7-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J8-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5J9-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5JA-01A-11R-A29S-07</th>\n",
       "      <th>TCGA-OR-A5JB-01A-11R-A29S-07</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-CG-4449-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4462-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4465-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4466-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4469-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4472-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4474-01A-02R-1157-13</th>\n",
       "      <th>TCGA-CG-4475-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4476-01A-01R-1157-13</th>\n",
       "      <th>TCGA-CG-4477-01A-01R-1157-13</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>?|100130426</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|100133144</th>\n",
       "      <td>3.2661</td>\n",
       "      <td>2.6815</td>\n",
       "      <td>1.7301</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.1673</td>\n",
       "      <td>1.4422</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>4.4556</td>\n",
       "      <td>7.1293</td>\n",
       "      <td>...</td>\n",
       "      <td>4.358154</td>\n",
       "      <td>5.676995</td>\n",
       "      <td>5.219350</td>\n",
       "      <td>14.846708</td>\n",
       "      <td>20.115492</td>\n",
       "      <td>6.997533</td>\n",
       "      <td>18.311906</td>\n",
       "      <td>12.057112</td>\n",
       "      <td>18.628740</td>\n",
       "      <td>17.874417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|100134869</th>\n",
       "      <td>3.9385</td>\n",
       "      <td>8.9948</td>\n",
       "      <td>6.5650</td>\n",
       "      <td>1.5492</td>\n",
       "      <td>4.4709</td>\n",
       "      <td>6.0529</td>\n",
       "      <td>2.2876</td>\n",
       "      <td>1.3599</td>\n",
       "      <td>5.0581</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.656360</td>\n",
       "      <td>3.342794</td>\n",
       "      <td>2.423442</td>\n",
       "      <td>5.055287</td>\n",
       "      <td>11.626054</td>\n",
       "      <td>13.654193</td>\n",
       "      <td>7.417109</td>\n",
       "      <td>11.585177</td>\n",
       "      <td>11.482418</td>\n",
       "      <td>14.919338</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|10357</th>\n",
       "      <td>149.1350</td>\n",
       "      <td>81.0777</td>\n",
       "      <td>86.4879</td>\n",
       "      <td>53.9117</td>\n",
       "      <td>66.9063</td>\n",
       "      <td>103.5060</td>\n",
       "      <td>94.9316</td>\n",
       "      <td>78.1955</td>\n",
       "      <td>69.2389</td>\n",
       "      <td>155.7090</td>\n",
       "      <td>...</td>\n",
       "      <td>633.299781</td>\n",
       "      <td>294.018042</td>\n",
       "      <td>686.569179</td>\n",
       "      <td>563.573453</td>\n",
       "      <td>1039.307597</td>\n",
       "      <td>639.238135</td>\n",
       "      <td>742.479964</td>\n",
       "      <td>506.336449</td>\n",
       "      <td>712.452165</td>\n",
       "      <td>703.713324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>?|10431</th>\n",
       "      <td>2034.1000</td>\n",
       "      <td>1304.9300</td>\n",
       "      <td>1054.6600</td>\n",
       "      <td>2350.8900</td>\n",
       "      <td>1257.9900</td>\n",
       "      <td>1866.4300</td>\n",
       "      <td>995.0270</td>\n",
       "      <td>1762.1200</td>\n",
       "      <td>1213.5300</td>\n",
       "      <td>2005.5700</td>\n",
       "      <td>...</td>\n",
       "      <td>1202.538277</td>\n",
       "      <td>644.002317</td>\n",
       "      <td>1181.884532</td>\n",
       "      <td>663.885074</td>\n",
       "      <td>647.530395</td>\n",
       "      <td>1297.152549</td>\n",
       "      <td>1152.909807</td>\n",
       "      <td>1375.495774</td>\n",
       "      <td>971.893874</td>\n",
       "      <td>1736.988111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 11069 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-OR-A5J1-01A-11R-A29S-07  TCGA-OR-A5J2-01A-11R-A29S-07  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                        0.0000                        0.0000   \n",
       "?|100133144                        3.2661                        2.6815   \n",
       "?|100134869                        3.9385                        8.9948   \n",
       "?|10357                          149.1350                       81.0777   \n",
       "?|10431                         2034.1000                     1304.9300   \n",
       "\n",
       "             TCGA-OR-A5J3-01A-11R-A29S-07  TCGA-OR-A5J5-01A-11R-A29S-07  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                        0.0000                        0.0000   \n",
       "?|100133144                        1.7301                        0.0000   \n",
       "?|100134869                        6.5650                        1.5492   \n",
       "?|10357                           86.4879                       53.9117   \n",
       "?|10431                         1054.6600                     2350.8900   \n",
       "\n",
       "             TCGA-OR-A5J6-01A-31R-A29S-07  TCGA-OR-A5J7-01A-11R-A29S-07  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                        0.0000                        0.0000   \n",
       "?|100133144                        0.0000                        1.1673   \n",
       "?|100134869                        4.4709                        6.0529   \n",
       "?|10357                           66.9063                      103.5060   \n",
       "?|10431                         1257.9900                     1866.4300   \n",
       "\n",
       "             TCGA-OR-A5J8-01A-11R-A29S-07  TCGA-OR-A5J9-01A-11R-A29S-07  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                        0.0000                        0.0000   \n",
       "?|100133144                        1.4422                        0.0000   \n",
       "?|100134869                        2.2876                        1.3599   \n",
       "?|10357                           94.9316                       78.1955   \n",
       "?|10431                          995.0270                     1762.1200   \n",
       "\n",
       "             TCGA-OR-A5JA-01A-11R-A29S-07  TCGA-OR-A5JB-01A-11R-A29S-07  ...  \\\n",
       "gene_id                                                                  ...   \n",
       "?|100130426                        0.0000                        0.0000  ...   \n",
       "?|100133144                        4.4556                        7.1293  ...   \n",
       "?|100134869                        5.0581                        0.0000  ...   \n",
       "?|10357                           69.2389                      155.7090  ...   \n",
       "?|10431                         1213.5300                     2005.5700  ...   \n",
       "\n",
       "             TCGA-CG-4449-01A-01R-1157-13  TCGA-CG-4462-01A-01R-1157-13  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                           NaN                           NaN   \n",
       "?|100133144                      4.358154                      5.676995   \n",
       "?|100134869                      2.656360                      3.342794   \n",
       "?|10357                        633.299781                    294.018042   \n",
       "?|10431                       1202.538277                    644.002317   \n",
       "\n",
       "             TCGA-CG-4465-01A-01R-1157-13  TCGA-CG-4466-01A-01R-1157-13  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                           NaN                           NaN   \n",
       "?|100133144                      5.219350                     14.846708   \n",
       "?|100134869                      2.423442                      5.055287   \n",
       "?|10357                        686.569179                    563.573453   \n",
       "?|10431                       1181.884532                    663.885074   \n",
       "\n",
       "             TCGA-CG-4469-01A-01R-1157-13  TCGA-CG-4472-01A-01R-1157-13  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                           NaN                           NaN   \n",
       "?|100133144                     20.115492                      6.997533   \n",
       "?|100134869                     11.626054                     13.654193   \n",
       "?|10357                       1039.307597                    639.238135   \n",
       "?|10431                        647.530395                   1297.152549   \n",
       "\n",
       "             TCGA-CG-4474-01A-02R-1157-13  TCGA-CG-4475-01A-01R-1157-13  \\\n",
       "gene_id                                                                   \n",
       "?|100130426                           NaN                           NaN   \n",
       "?|100133144                     18.311906                     12.057112   \n",
       "?|100134869                      7.417109                     11.585177   \n",
       "?|10357                        742.479964                    506.336449   \n",
       "?|10431                       1152.909807                   1375.495774   \n",
       "\n",
       "             TCGA-CG-4476-01A-01R-1157-13  TCGA-CG-4477-01A-01R-1157-13  \n",
       "gene_id                                                                  \n",
       "?|100130426                           NaN                           NaN  \n",
       "?|100133144                     18.628740                     17.874417  \n",
       "?|100134869                     11.482418                     14.919338  \n",
       "?|10357                        712.452165                    703.713324  \n",
       "?|10431                        971.893874                   1736.988111  \n",
       "\n",
       "[5 rows x 11069 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_expr_df = pd.read_csv(\n",
    "    os.path.join(cfg.raw_data_dir, manifest_df.loc['rna_seq'].filename),\n",
    "    index_col=0, sep='\\t')\n",
    "\n",
    "print(tcga_expr_df.shape)\n",
    "tcga_expr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Process gene expression matrix\n",
    "\n",
    "This involves updating Entrez gene ids, sorting and subsetting."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set index as entrez_gene_id\n",
    "tcga_expr_df.index = tcga_expr_df.index.map(lambda x: x.split('|')[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "tcga_expr_df = (tcga_expr_df\n",
    "    .dropna(axis='rows')\n",
    "    .rename(index=old_to_new_entrez)\n",
    "    .groupby(level=0).mean()\n",
    "    .transpose()\n",
    "    .sort_index(axis='rows')\n",
    "    .sort_index(axis='columns')\n",
    ")\n",
    "\n",
    "tcga_expr_df.index.rename('sample_id', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update sample IDs to remove multiple samples measured on the same tumor\n",
    "# and to map with the clinical information\n",
    "tcga_expr_df.index = tcga_expr_df.index.str.slice(start=0, stop=15)\n",
    "tcga_expr_df = tcga_expr_df.loc[~tcga_expr_df.index.duplicated(), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for valid Entrez gene identifiers\n",
    "tcga_expr_df = tcga_expr_df.loc[:, tcga_expr_df.columns.isin(gene_df.entrez_gene_id.astype(str))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "expr_file = os.path.join(cfg.data_dir, 'tcga_expression_matrix_processed.tsv.gz')\n",
    "tcga_expr_df.to_csv(expr_file, sep='\\t', compression='gzip', float_format='%.3g')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(tcga_expr_df.shape)\n",
    "tcga_expr_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Process TCGA cancer-type and sample-type info from barcodes\n",
    "\n",
    "Cancer-type includes `OV`, `BRCA`, `LUSC`, `LUAD`, etc. while sample-type includes `Primary`, `Metastatic`, `Solid Tissue Normal`, etc.\n",
    "\n",
    "See https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/tissue-source-site-codes for more details.\n",
    "\n",
    "The goal is to use this info to stratify training (90%) and testing (10%) balanced by cancer-type and sample-type. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract sample type in the order of the gene expression matrix\n",
    "tcga_id = pd.DataFrame(tcga_expr_df.index)\n",
    "\n",
    "# Extract the last two digits of the barcode and recode sample-type\n",
    "tcga_id = tcga_id.assign(sample_type = tcga_id.sample_id.str[-2:])\n",
    "tcga_id.sample_type = tcga_id.sample_type.replace(sampletype_codes_dict)\n",
    "\n",
    "# Extract the first two ID numbers after `TCGA-` and recode cancer-type\n",
    "tcga_id = tcga_id.assign(cancer_type = tcga_id.sample_id.str[5:7])\n",
    "tcga_id.cancer_type = tcga_id.cancer_type.replace(cancertype_codes_dict)\n",
    "\n",
    "# Append cancer-type with sample-type to generate stratification variable\n",
    "tcga_id = tcga_id.assign(id_for_stratification = tcga_id.cancer_type.str.cat(tcga_id.sample_type))\n",
    "\n",
    "# Get stratification counts - function cannot work with singleton strats\n",
    "stratify_counts = tcga_id.id_for_stratification.value_counts().to_dict()\n",
    "\n",
    "# Recode stratification variables if they are singletons\n",
    "tcga_id = tcga_id.assign(stratify_samples_count = tcga_id.id_for_stratification)\n",
    "tcga_id.stratify_samples_count = tcga_id.stratify_samples_count.replace(stratify_counts)\n",
    "tcga_id.loc[tcga_id.stratify_samples_count == 1, \"stratify_samples\"] = \"other\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write out files for downstream use\n",
    "file = os.path.join(cfg.data_dir, 'tcga_sample_identifiers.tsv')\n",
    "\n",
    "(\n",
    "    tcga_id.drop(['stratify_samples', 'stratify_samples_count'], axis='columns')\n",
    "    .to_csv(file, sep='\\t', index=False)\n",
    ")\n",
    "\n",
    "print(tcga_id.shape)\n",
    "tcga_id.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cancertype_count_df = (\n",
    "    pd.DataFrame(tcga_id.cancer_type.value_counts())\n",
    "    .reset_index()\n",
    "    .rename({'index': 'cancertype', 'cancer_type': 'n ='}, axis='columns')\n",
    ")\n",
    "\n",
    "file = os.path.join(cfg.data_dir, 'tcga_sample_counts.tsv')\n",
    "cancertype_count_df.to_csv(file, sep='\\t', index=False)\n",
    "\n",
    "cancertype_count_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:mpmp]",
   "language": "python",
   "name": "conda-env-mpmp-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
